### PR TITLE
clustermesh: add a readme explaining MCS-API implementation

### PR DESCRIPTION
**Pull request by MrFreezeex**
Add a short readme explaining the different components involved in the MCS-API implementation with an associated mermaid schema.

Followup to https://github.com/cilium/cilium/pull/34439#discussion_r1794891020

### CODE REVIEW

**Comment by MrFreezeex** [comment]
Hi @squeed, as promised here is the small readme with some more high level bits that should explains how everything is tied together, let me know if that's what you had in mind :pray:.

File: pkg/clustermesh/mcsapi/README.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+                namespace: my-ns
+            spec: (...)
+        `"]
+        remoteServiceExport["`
+            ServiceExport
+            metadata:
+                name: my-svc
+                namespace: my-ns
+        `"]
+        clustermeshAPIServer@{ shape: cyl, label: "ClusterMesh API Server" }
```
Comment (giorio94) [inline pkg/clustermesh/mcsapi/README.md]: The GH preview doesn't seem to like this line.

**Review by giorio94 — COMMENTED** [review]
Thanks!

I'm personally not a big fan of this type of README files, as they tend to get stale over time given that no one ever remembers to update them subsequently. However, I don't actually see any better place for this type of information --- there's apparently no good location in the docs at the moment, and a `doc.go` file would suffer the same limitations (the only advantage is that it would be automatically used by godoc). 

Content wise, my main suggestion would be to reorganize a bit the content, as I feel that in the current form may be a bit difficult to understand without the full context. One possible idea could be:

* Adding a brief introductory paragraph with a bit more context, as simple as mentioning what MCS-API is and linking to the KEP;
* Having a bullet list (or a couple of paragraphs) about the core types, simply linking to the dedicated KEP sections when appropriate.
* Having a bullet list (or subsections) with one item for every controller (maybe mentioning in which file they are implemented?), briefly describing what they do, which are the inputs and outputs;
 * Briefly mentioning what conflict resolution is, and which role plays here -- again, just a simple sentence linking to the KEP, and possibly any additional info for non-trivial implementation choices where the KEP was not clear.

WDYT?

**Comment by MrFreezeex** [comment]
> I'm personally not a big fan of this type of README files, as they tend to get stale over time given that no one ever remembers to update them subsequently. However, I don't actually see any better place for this type of information --- there's apparently no good location in the docs at the moment, and a `doc.go` file would suffer the same limitations (the only advantage is that it would be automatically used by godoc).

Yeah I agree, I am trying to be a bit high level here so that it hopefully decrease the change of being out of date but this might still happen :sweat_smile:.

> Content wise, my main suggestion would be to reorganize a bit the content, as I feel that in the current form may be a bit difficult to understand without the full context. One possible idea could be:
> 
>     * Adding a brief introductory paragraph with a bit more context, as simple as mentioning what MCS-API is and linking to the KEP;
>     * Having a bullet list (or a couple of paragraphs) about the core types, simply linking to the dedicated KEP sections when appropriate.

I tried explaining this in a small intro let me know if that's what you had in mind!

>     * Having a bullet list (or subsections) with one item for every controller (maybe mentioning in which file they are implemented?), briefly describing what they do, which are the inputs and outputs;

Hmm this is already sort of done in the paragraphs before the schema, it's not in a super structured way though. My intent was more to describe the general flow and the controller involved than independently what each controllers there is also some documentation on each controller about what they individually do :man_shrugging:.

>     * Briefly mentioning what conflict resolution is, and which role plays here -- again, just a simple sentence linking to the KEP, and possibly any additional info for non-trivial implementation choices where the KEP was not clear.

I added a small note about that, also hopefully the KEP is going to be updated with what was not clear on the ServiceImport PR anyway.

File: pkg/clustermesh/mcsapi/README.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
+# Multi-Cluster Services API
+
+Multi-Cluster Services API (MCS-API) is a standard API extending the Service concept
+in multiple clusters. For more information about MCS-API see the [KEP-1645](https://github.com/kubernetes/enhancements/blob/master/keps/sig-multicluster/1645-multi-cluster-services-api/README.md).
+
+MCS-API relies on two new CRDs: ServiceExport and ServiceImport. A ServiceExport
+is a simple resource marking a Service of the same name and namespace for
+exporting it into other clusters. The ServiceImport resource is entirely managed
+by its implementation (by a Cilium controller in our case) and represent the
```
Comment (giorio94) [inline pkg/clustermesh/mcsapi/README.md]:
```suggestion
by its implementation (by a Cilium controller in our case) and represents the
```

File: pkg/clustermesh/mcsapi/README.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+The implementation relies on a dedicated struct in the `types` subpackage named `MCSAPIServiceSpec`
+which contains the different fields of the exported services (a Service and a ServiceExport of the same name).
+This struct is synchronized to the Clustermesh API Server like other clustermesh structs
+not directly tied to MCS-API. It is then pulled from all the clusters to reconstruct the
+ServiceImport resources by the controller `mcsAPIServiceImportReconciler`.
+
+Note that exporting a Service can result in a conflict, [see the dedicated
+section in the KEP describing conflicts](https://github.com/kubernetes/enhancements/blob/master/keps/sig-multicluster/1645-multi-cluster-services-api/README.md#constraints-and-conflict-resolution).
+Also unlike some other implementations we handle conflict at import time (even though
```
Comment (giorio94) [inline pkg/clustermesh/mcsapi/README.md]:
```suggestion
Also unlike some other implementations we handle conflicts at import time (even though
```

File: pkg/clustermesh/mcsapi/README.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+this information is exposed to the user on the ServiceExport object). That's because the Cilium
+way of exporting a Service is to update its local clustermesh API Server while at import time
+it aggregates the information from every clustermesh API Server. Thus this is also handled
+when we construct the ServiceImport resource in the `mcsAPIServiceImportReconciler` controller.
+
+Then based on this ServiceImport another controller `mcsAPIServiceReconciler` will
+create an actual Service named `derived-$hash` which is internal Service designed to
```
Comment (giorio94) [inline pkg/clustermesh/mcsapi/README.md]:
```suggestion
create an actual Service named `derived-$hash` which is the internal Service designed to
```

File: pkg/clustermesh/mcsapi/README.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+trigger the normal mechanism of a regular Service. This involves for instance ClusterIP
```
Comment (giorio94) [inline pkg/clustermesh/mcsapi/README.md]:
```suggestion
trigger the normal mechanism of a regular global Service. This involves for instance ClusterIP
```

**Review by giorio94 — APPROVED** [review]
Looks great, thanks!

**Pull request merged**